File: benchmarks/baseline_bn_par.f90
module bnmod
  implicit none
contains
  subroutine process_perm(perm, n, hist, nxt, visited)
    integer, intent(in) :: perm(:), n
    integer(8), intent(inout) :: hist(:)
    integer, intent(inout) :: nxt(:), visited(:)
    integer :: twoN, limit, mask, m, i, bit, v, r, cycles, s, cur
    twoN = 2*n
    limit = 1
    do i = 1, n
       limit = limit * 2
    end do
    do mask = 0, limit-1
       m = mask
       do i = 1, n
          bit = m - (m/2)*2
          v = perm(i)
          if (bit == 0) then
             nxt(i) = v
             nxt(n+i) = n+v
          else
             nxt(i) = n+v
             nxt(n+i) = v
          end if
          m = m/2
       end do
       do r = 1, twoN
          visited(r) = 0
       end do
       cycles = 0
       do s = 1, twoN
          if (visited(s) == 0) then
             cycles = cycles + 1
             cur = s
             do while (visited(cur) == 0)
                visited(cur) = 1
                cur = nxt(cur)
             end do
          end if
       end do
       hist(cycles) = hist(cycles) + 1
    end do
  end subroutine process_perm

  recursive subroutine heapsub(m, base, perm, n, hist, nxt, visited)
    integer, intent(in) :: m, base, n
    integer, intent(inout) :: perm(:), nxt(:), visited(:)
    integer(8), intent(inout) :: hist(:)
    integer :: i, t
    if (m == 1) then
       call process_perm(perm, n, hist, nxt, visited)
       return
    end if
    do i = 1, m-1
       call heapsub(m-1, base, perm, n, hist, nxt, visited)
       if (mod(m,2) == 0) then
          t = perm(base+i-1); perm(base+i-1) = perm(base+m-1); perm(base+m-1) = t
       else
          t = perm(base); perm(base) = perm(base+m-1); perm(base+m-1) = t
       end if
    end do
    call heapsub(m-1, base, perm, n, hist, nxt, visited)
  end subroutine heapsub

  function compute_prefix(first, n) result(cnt)
    integer, intent(in) :: first, n
    integer(8) :: cnt
    integer, allocatable :: perm(:), nxt(:), visited(:)
    integer(8), allocatable :: hist(:)
    integer :: twoN, i, v, idx
    twoN = 2*n
    allocate(perm(n), nxt(twoN), visited(twoN), hist(twoN))
    perm(1) = first
    idx = 2
    do v = 1, n
       if (v /= first) then
          perm(idx) = v
          idx = idx + 1
       end if
    end do
    hist = 0
    call heapsub(n-1, 2, perm, n, hist, nxt, visited)
    cnt = 0
    do i = 1, twoN
       cnt = cnt + hist(i)
    end do
    deallocate(perm, nxt, visited, hist)
  end function compute_prefix
end module bnmod

program bn_par
  use bnmod
  implicit none
  integer :: n, first, nargs
  integer(8) :: total
  character(len=32) :: arg
  n = 10
  nargs = command_argument_count()
  if (nargs >= 1) then
     call get_command_argument(1, arg)
     read(arg, *) n
  end if
  total = 0
  !$omp parallel do reduction(+:total) schedule(dynamic)
  do first = 1, n
     total = total + compute_prefix(first, n)
  end do
  !$omp end parallel do
  print '(A,I0)', 'total=', total
end program bn_par

File: benchmarks/baseline_bn_par.c
#include <stdio.h>
#include <stdlib.h>
static int N, twoN;
static void process_perm(int *perm, long long *hist, int *nxt, int *visited){
    int limit=1,b; for(b=0;b<N;b++) limit*=2;
    int mask;
    for(mask=0; mask<limit; mask++){
        int m=mask,i;
        for(i=0;i<N;i++){
            int bit=m-(m/2)*2, v=perm[i];
            if(bit==0){ nxt[i]=v-1; nxt[N+i]=N+v-1; }
            if(bit==1){ nxt[i]=N+v-1; nxt[N+i]=v-1; }
            m/=2;
        }
        int r; for(r=0;r<twoN;r++) visited[r]=0;
        int cycles=0,s;
        for(s=0;s<twoN;s++){
            if(!visited[s]){ cycles++; int cur=s; while(!visited[cur]){visited[cur]=1;cur=nxt[cur];} }
        }
        hist[cycles]++;
    }
}
static void heapsub(int m,int base,int *perm,long long *hist,int *nxt,int *visited){
    if(m==1){ process_perm(perm,hist,nxt,visited); return; }
    int i;
    for(i=0;i<m-1;i++){
        heapsub(m-1,base,perm,hist,nxt,visited);
        if((m-(m/2)*2)==0){int t=perm[base+i];perm[base+i]=perm[base+m-1];perm[base+m-1]=t;}
        if((m-(m/2)*2)==1){int t=perm[base];perm[base]=perm[base+m-1];perm[base+m-1]=t;}
    }
    heapsub(m-1,base,perm,hist,nxt,visited);
}
static long long compute_prefix(int first){
    int perm[20], nxt[40], visited[40]; long long hist[41];
    perm[0]=first; int idx=1,v;
    for(v=1;v<=N;v++) if(v!=first) perm[idx++]=v;
    int c; for(c=0;c<twoN+1;c++) hist[c]=0;
    heapsub(N-1,1,perm,hist,nxt,visited);
    long long cnt=0; for(c=0;c<twoN+1;c++) cnt+=hist[c];
    return cnt;
}
int main(int argc,char**argv){
    N=argc>1?atoi(argv[1]):10; twoN=2*N;
    long long total=0; int first;
    #pragma omp parallel for reduction(+:total) schedule(dynamic)
    for(first=1;first<=N;first++) total += compute_prefix(first);
    printf("total=%lld\n",total);
    return 0;
}
